File: 257-BinaryTreePaths/257BinaryTreePaths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 vector<string> res;
 vector<int> path;
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        res.clear();
        preOrder(root);
        return res;
    }
    string intVectorToString(vector<int> & v){
        string res;
        for(auto i : v){
            res += to_string(i) + "->";
        }
        if(res.size()) return string(res.begin(), res.end()-2);
        return res;
    }
    void preOrder(TreeNode* root){
        if(root){
            path.push_back(root->val);
            if(root->left) preOrder(root->left);
            // bad implementations, leave for future
            if((!root->left) && (!root->right)) res.push_back(intVectorToString(path));
            if(root->right) preOrder(root->right);
            path.pop_back();
        }
    }
};

File: 257-BinaryTreePaths/257. Binary Tree Paths.py
__author__ = 'liuxiyun'

# Backtracking or DFS
# Time: O(n) = # of path = n/2
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        self.res = []
        if root==None:
            return []
        temp =''
        node=root
        self.dfs(temp,node)
        return self.res
    def dfs(self,temp,node):
        if node.left==None and node.right==None:
            self.res.append(temp+str(node.val))
            return
        if node.left!=None:
            self.dfs(temp+str(node.val)+'->',node.left)
        if node.right!=None:
            self.dfs(temp+str(node.val)+'->',node.right)

# Test case
# # []
# # [1,2,3]
# # [1,2,3,#,#,4,#]

